### PR TITLE
Reassigns a few conflicting radio channel keys (department, entertainment, deadsay), adds unit test for preventing such conflicts, and fixes a minor issue in error handling

### PR DESCRIPTION
**Pull request by vinylspiders**
## About The Pull Request

This PR adds insurance that every radio/saymode key is unique, complete with a unit test. Providing detailed messages for why CI is failing when the user tries to add one that is already used by -something-.

Fixes a few conflicts with radio keys that were found with these tests. See changelog for the ones that have been changed. Notably department is no longer ":h" but ":d" (makes sense right?)

---

Encountering a radio channel duplicate is a very annoying experience currently, as they were very difficult to track across the various global lists. This should make that process far less painful.

### Notable features:

<details><summary>Runtimes when trying to add an existing key instead of just silently overwriting the previous key's value</summary>

dreamseeker_BJ6MNQbC3j

</details>

<details><summary>CI Shown here</summary>

Code_478ZCZ1lqJ

Code_NP1NWkVynb

</details>

Amusingly enough in my testing I also found an error with the stack_trace() proc itself, where apparently if the runtime happens early enough you get this runtime here:

<details><summary>this was preventing the runtime from displaying in the log</summary>

dreamdaemon_gwnGfX9BaO

</details>

Fixed that too by making sure that static regex gets initialized.

## Why It's Good For The Game

Preventing channel conflicts is now a lot easier. Support for uppercase keys is kind of necessary as we are running out of letters, and will help ease the limitations there for future expansions.

<details><summary>lowercase/uppercase get treated the same if there is no uppercase key for that letter</summary>

u90PpXrHoM

</details>

## Changelog

:cl:
qol: added support for uppercase radio channel keys. if there doesn't exist an uppercased key for that letter, the key will be automatically lowercased as it was previously
fix: department radio channel key has been swapped from ":h" to ":d" in order to fix a conflict with the holopad key
fix: entertainment radio channel key has been swapped from ":p" to ":K" in order to fix a conflict with the asay key
fix: deadsay channel key has been swapped from ":d" to ":J" in order to fix conflict with the department key (don't think this one gets used ever)
fix: (cyrillic) department radio channel key has been swapped from ":р" to ":в" for parity with the roman keyboard layout
fix: (cyrillic) entertainment radio channel key has been swapped from ":з" to ":ж" to fix a conflict with the asay key
fix: (cyrillic) deadsay channel key has been swapped from ":в" to ":д" in order to fix conflict with the department key
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
Important note, someone did this a while back and it was reverted in #71474 

Some keys are (apparently) supposed to conflict, and having a test that disallows you from having conflicting keys stops that

**Comment by necromanceranne** [comment]
It is still fucking ridiculous that this continues to be an ongoing problem that we repeatedly have devs stumble into that clearly can be easily solved, but cant be because the fucking server will get forked. Good god.

File: code/__DEFINES/radio.dm [hunk 1]
```diff
@@ -38,9 +38,8 @@
 #define RADIO_TOKEN_AI_PRIVATE ":o"
 
 #define RADIO_CHANNEL_ENTERTAINMENT "Entertainment"
-#define RADIO_KEY_ENTERTAINMENT "p"
-#define RADIO_TOKEN_ENTERTAINMENT ":p"
-
+#define RADIO_KEY_ENTERTAINMENT "K"
+#define RADIO_TOKEN_ENTERTAINMENT ":K"
```
Comment (LemonInTheDark) [inline code/__DEFINES/radio.dm:42]: capital K?

File: code/__DEFINES/say.dm [hunk 2]
```diff
@@ -31,14 +31,14 @@
 #define MODE_WHISPER_CRIT "whispercrit"
 
 #define MODE_DEPARTMENT "department"
-#define MODE_KEY_DEPARTMENT "h"
-#define MODE_TOKEN_DEPARTMENT ":h"
+#define MODE_KEY_DEPARTMENT "d"
+#define MODE_TOKEN_DEPARTMENT ":d"
```
Comment (LemonInTheDark) [inline code/__DEFINES/say.dm:35]: I'm fine with this but you're gonna ruin a lot of people's muscle memories

File: code/__DEFINES/say.dm [hunk 3]
```diff
@@ -31,14 +31,14 @@
...
 
 #define MODE_ADMIN "admin"
 #define MODE_KEY_ADMIN "p"
 
 #define MODE_DEADMIN "deadmin"
-#define MODE_KEY_DEADMIN "d"
+#define MODE_KEY_DEADMIN "J"
```
Comment (LemonInTheDark) [inline code/__DEFINES/say.dm:41]: why j also why ccapitalized that's dumb

File: code/modules/error_handler/error_handler.dm [hunk 4]
```diff
@@ -26,7 +26,9 @@ GLOBAL_VAR_INIT(total_runtimes_skipped, 0)
 		Reboot(reason = 1)
 		return
 
-	var/static/regex/stack_workaround = regex("[WORKAROUND_IDENTIFIER](.+?)[WORKAROUND_IDENTIFIER]")
+	var/static/regex/stack_workaround
+	if(isnull(stack_workaround))
+		stack_workaround = regex("[WORKAROUND_IDENTIFIER](.+?)[WORKAROUND_IDENTIFIER]")
```
Comment (LemonInTheDark) [inline code/modules/error_handler/error_handler.dm:31]: I hate order of operations

File: code/modules/mob/dead/observer/observer_say.dm [hunk 5]
```diff
@@ -5,8 +5,12 @@
 /mob/dead/observer/get_message_mods(message, list/mods)
 	var/key = message[1]
 	if((key in GLOB.department_radio_prefixes) && length(message) > length(key) + 1 && !mods[RADIO_EXTENSION])
-		mods[RADIO_KEY] = LOWER_TEXT(message[1 + length(key)])
-		mods[RADIO_EXTENSION] = GLOB.department_radio_keys[mods[RADIO_KEY]]
+		mods[RADIO_KEY] = message[1 + length(key)]
+		// intelligently use lowercase if there doesn't exist an uppercase version of that letter in department_radio_keys
```
Comment (LemonInTheDark) [inline code/modules/mob/dead/observer/observer_say.dm:9]: I don't think we should have uppercase/lowercase keys people already don't remember these

File: code/modules/mob/dead/observer/observer_say.dm [hunk 6]
```diff
@@ -5,8 +5,12 @@
...
+		var/radio_extension = GLOB.department_radio_keys[mods[RADIO_KEY]]
+		if(isnull(radio_extension))
```
Comment (LemonInTheDark) [inline code/modules/mob/dead/observer/observer_say.dm:11]: this code is mildly hot make the other list set an else here

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
gonna poke mso bout this

File: code/__DEFINES/say.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/__DEFINES/say.dm:35]: the fact that someone just accidentally shadowed one is a good indicated that we need to do something I agree

File: code/__DEFINES/say.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/__DEFINES/say.dm:41]: from a quick admin poll I don't think any of them know about puppet either (I'm gonna remove it.)
that said I have opinions about the renames so uh, please hold

**Comment by MrStonedOne** [comment]
I have reverted this exact pr before.

**I would prefer keys be allowed to conflict in different contexts that don't occur at the same time.**

(as an aside: i almost think it might be better if we remove the asay radio key because that's putting two very different streams too close together. (for example, if you use it while forgetting you are de-admined, does it eat the message or show everybody what you meant only admins to see? ), like that on its own is not enough, but we are running out of radio channel keys.)

even on top of that, as far as im concern, the holopad **is** the departmental radio when an ai is in a holopad. that doesn't have to be viewed as a conflict if you aren't needlessly rigid about it. Even under your change you would need to preserve that functionality so that :d goes to the holopad when an ai is in a holopad because :h/:d is suppose to be a convenience shortcut that goes to the most obvious place and an ai in a holopad would expect department radio (:d) to go to the holopad they are in.

Every time this comes up i bring up the same point.

**We are running out of radio channel keys** We can not afford to reserve them for one use in all contexts. Allowing them to conflict in theory but not in practice like with alien hivemind or other limited context uses is a smarter way of allocating the limited resource.

Moving into upper case keys or letters is a huge cost that is only justified by trying to write some easy to blanket enforce rule rather then spending the time to do it right.

For instances: In starcraft 2, it is impossibleto bind every ability to a key, so the keybinding system allows for some types of conflicts and resolves them in the field based on a priority, where as other kinds of keybinding conflicts are not allowed, like (some) ui actions, without modifying the config file yourself. (also, what the mouse does when you left click is a key bind that has 9 conflicts by default that some progamers rebind to space bar)

We need to embrace nuance like this to make better use of the limited space that exists for radio keys.

**Comment by LemonInTheDark** [comment]
> (Other thing, is I know some downstreams have a lot of extra radio channels, and it might be a good idea to keep the -support- of uppercased keys for the niche ones, but we don't necessarily have to make use of them here)

IDK. It's not good ux so I would prefer to not make the option open to our coders, and if we aren't using it then it's just kinda dead code to use right?

File: code/__DEFINES/say.dm (hunk 3 above)
Comment (Ghommie) [reply code/__DEFINES/say.dm:41]: > from a quick admin poll I don't think any of them know about puppet either (I'm gonna remove it.) that said I have opinions about the renames so uh, please hold

So did you remove it?!?

Comment (LemonInTheDark) [reply code/__DEFINES/say.dm:41]: I uh, did not no

Comment (MrMelbert) [reply code/__DEFINES/say.dm:41]: WAIT NOOO DON'T REMOVE PUPPET IT'S BASED

Comment (MrMelbert) [reply code/__DEFINES/say.dm:41]: We just need to advertise its existence better (I use it literally every round I admin)

Comment (LemonInTheDark) [reply code/__DEFINES/say.dm:41]: it should be a verb instead of a chat prefix it's confusing and not made clear to jannies

Comment (MrMelbert) [reply code/__DEFINES/say.dm:41]: As long as it exists in some form then i'm chill

Comment (MrStonedOne) [reply code/__DEFINES/say.dm:41]: ya that seemed fine to me, sorry if i forgot to say such.

Comment (LemonInTheDark) [reply code/__DEFINES/say.dm:41]: looks good to me

**Comment by LemonInTheDark** [comment]
sadge

**Pull request closed without merging**